File: Data Structures & Algorithms/n-th-tribonacci-number/submission-3.cpp
class Solution {
public:
    int tribonacci(int n) {
                if (n == 0)
        {
            return 0;
        }
        if (n == 1 || n == 2)
        {
            return 1;
        }

        vector<int> tri(n + 1, 0);
        tri[1] = 1;
        tri[2] = 1;
        for (int index = 3; index < n + 1; index++)
        {
            tri[index] = tri[index - 1] + tri[index - 2] + tri[index - 3];
        }
        return tri[n];
    }
};

File: Data Structures & Algorithms/minimum-height-trees/submission-3.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        adj = [[] for _ in range(n)]
        for source, destination in edges:
            adj[source].append(destination)
            adj[destination].append(source)

        degree = [(len(adj[i])) for i in range(n)]
        leaves = [index for index in range(n) if degree[index] == 1]
        remaining = n
        while remaining > 2:
            remaining -= len(leaves)
            newLeaves = []

            for leaf in leaves:
                for neighbor in adj[leaf]:
                    degree[neighbor] -= 1
                    if degree[neighbor] == 1:
                        newLeaves.append(neighbor)
            leaves = newLeaves

        return leaves